### PR TITLE
feat: use engine implementation of findFirstOrThrow / findUniqueOrThrow

### LINKED ISSUE #10642: When rejectOnNotFound is used, chaining deeper into a related table still throws if it doesn't find anything
### Bug description

`rejectOnNotFound` should not cause chained queries to fail if they don't find an entity.
It is best understood by looking at the reproduction example:

### How to reproduce

Have a code like this:
```ts
import { PrismaClient, Prisma } from '@prisma/client'

const prismaClient = new PrismaClient()

; (async () => {
        const existingUser = await prismaClient.user.findUnique({
            where: {
                id: 1
            },
        })
        console.log('~ existingUser', existingUser)
        const user = prismaClient.user.findUnique({
            where: {
                id: 1
            },
            rejectOnNotFound: true
        })
        console.log( await user.profilePicture())
})()
```

Running this code using `ts-node pr.ts` prints out:
```
~ existingUser {
  id: 1,
  createdAt: 2021-12-10T11:24:30.096Z,
  email: 'john@example.com',
  name: null,
  role: 'USER',
  profilePictureId: null
}
[NotFoundError: No User found] { clientVersion: '3.6.0' }
```

### Expected behavior

Expected output for me is this:
```
~ existingUser {
  id: 1,
  createdAt: 2021-12-10T11:24:30.096Z,
  email: 'john@example.com',
  name: null,
  role: 'USER',
  profilePictureId: null
}
null
```
Without throwing any errors.

### Prisma information
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./bugRepro.db"
}

model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  email     String   @unique
  name      String?

  role             String @default("USER")
  profilePictureId Int?
  profilePicture   Image? @relation(fields: [profilePictureId], references: [id])
}

model Image {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  isHeader  Boolean  @default(false)
  url       String
  User      User[]
}

```
### Environment & setup

- OS: ubuntu
- Database: any
- Node.js version: 17

### Prisma Version

```
prisma                  : 3.6.0
@prisma/client          : 3.6.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : dc520b92b1ebb2d28dc3161f9f82e875bd35d727
Studio                  : 0.440.0

```

### PR DESCRIPTION
**Pull request by miguelff**
Fixes https://github.com/prisma/prisma/issues/10642
Closes https://github.com/prisma/prisma/pull/15887

Client changes to use [the engine implementation of `findFirstOrThrow` / `findUniqueOrThrow`](https://github.com/prisma/prisma-engines/pull/3356) rather than the emulated client actions added in https://github.com/prisma/prisma/pull/13920.

Once this is merged, I will tackle the following: 
- [x] https://github.com/prisma/prisma/issues/14933

### CODE REVIEW

File: packages/client/src/runtime/RequestHandler.ts
Comment (millsp) [inline packages/client/src/runtime/RequestHandler.ts:193]: Will this be made a PrismaClientKnownRequestError in the future?

Comment (millsp) [reply packages/client/src/runtime/RequestHandler.ts:193]: Found the answer here https://github.com/prisma/prisma/pull/16162/files#diff-99659f0376b81639e19aee7c9cf5dee20ca6f2a4d68cdf549b2439940466d5f3R14

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (millsp) [inline packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:2723]: I think these args a really the same as `PostFindFirstArgs`, could we maybe not emit these or is it too difficult?

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (millsp) [inline packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:4039]: Question: Are `UserFindUniqueArgsBase` still being generated? If so could we remove them?

File: packages/client/src/runtime/core/model/applyOrThrowErrorAdapter.ts
Comment (millsp) [inline packages/client/src/runtime/core/model/applyOrThrowErrorAdapter.ts:15]:
```suggestion
 * If the above mentioned methods returned no result, an error of type NotFoundError
```

File: packages/client/src/runtime/core/model/applyOrThrowErrorAdapter.ts
Comment (millsp) [inline packages/client/src/runtime/core/model/applyOrThrowErrorAdapter.ts:18]:
```suggestion
 * error thrown by the Query Engine and rethrow it as a client side error.
```

**Review by millsp — COMMENTED** [review]
I think the PR is in a good state, good work on this non-trivial implementation! :tada: I have left some comments/questions to see if we can improve things a bit. Will approve once resolved :)

File: packages/client/src/runtime/RequestHandler.ts
Comment (miguelff) [reply packages/client/src/runtime/RequestHandler.ts:193]: Yes, that's the intention. For the moment the NotFoundError became deprecated, it inherited from `PrismaClientKnownRequestError` thus getting a code and the rest of its properties, and once we reach the next major version it should be replaced.

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (miguelff) [reply packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap]: Good call, I did the code changes in https://github.com/prisma/prisma/pull/16162/commits/1b806115f3619bbfc6cedf5e961afe4e704d17ce and pushed the updated snapshots in https://github.com/prisma/prisma/pull/16162/commits/b6decd849ccf0a02778283813ecf10968d661f65

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (miguelff) [reply packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:4039]: They are being generated and used for [extending the type to deprecate `rejectOnNotFound`](https://github.com/prisma/prisma/blob/b6decd849ccf0a02778283813ecf10968d661f65/packages/client/src/generation/TSClient/Args.ts#L111-L123)

I'm not completely sure how to monomorphise the base type and remove the child definition, as it's the only one. I'm going to dig a little bit and see what I find.

Comment (miguelff) [reply packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:4039]: Now tests are breaking because `ModelFind(First|Unique)Args` are duplicated. They  are generated once from the original type, and then a second time add the rejectOnNotFound option. 

We can do two things in here:
1. We could generate only one type, having the optional `rejectOnNotFound` property for both `Find(First|Unique)`  `Find(First|Unique)OrThrow` 
2. We could revert the changes and leave the aliased type

My understanding is that 1 will introduce an unexpected change. Before this, findUnique or findFirst, wouldn't allow `rejectOnNotFound` anymore.  It's only OrThrow variants the ones that do, by doing 1, we will be promoting to all the input types the rejectOnNotFound property. -- Is that maybe a bug? in any case it doesn't make sense to take back a backward compatibility that once was broken --

Comment (miguelff) [reply packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:4039]: Given the above, I think the best course of action as of now is to do 2. and revert the changes and leave the aliased type. And I did that in https://github.com/prisma/prisma/pull/16162/commits/3f9eedb0934050886d9403149a5c2cf7ebc11a04 and https://github.com/prisma/prisma/pull/16162/commits/c55b6e74f71f6293bfa77c78db021bc335fe2512

I will remove the revert and reverted commits before merge.

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (SevInf) [inline packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:2195]: Might be a good idea to update jsdoc to refer to the error code instead of deprecated class. Not critical though, at least not until class gets removed.

**Review by SevInf — COMMENTED** [review]
One (optional) comment in addition to what Pierre said.
Great job! This certainly wasn't a trivial removal.

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap
Comment (miguelff) [reply packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap]: 👍🏾 Addressed in [64b0f70](https://github.com/prisma/prisma/pull/16162/commits/64b0f7057b3c03929a5c13bc74588381ec8f9a7d)

**Review by SevInf — APPROVED** [review]

**Pull request merged**
